### PR TITLE
Feature/merge bootstrap with extraction

### PR DESCRIPTION
**Pull request by njordan-hsr**
Merge bootstrapping and extracting into one container and make it work again similarly as before.

Implements parts of [#243](https://github.com/geometalab/osmaxx/issues/243)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,28 @@
...
+include/
+lib/
+local/
+
+# Python 3 centralized *.pyc location
+__pycache__/
+
+# Hidden by naming convention, but tracked anyway:
+!.gitignore
+!.travis.yml
```
Comment (das-g) [inline .gitignore]: Should pattern in this file be limited to the repository root? (E.g. ignore `<repo root>/html/index.html`, but not `<repo root>/some/sub/dirs/html/index.html`)

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+sudo: false
+
+env:
+    - TOX_ENV=py34-flake8
+    - TOX_ENV=py34-docs
+    - TOX_ENV=py33-django1.8-drf3.2
+    - TOX_ENV=py34-django1.8-drf3.2
+
+matrix:
+  fast_finish: true
```
Comment (das-g) [inline .travis.yml]: Not using [finishing fast in case of failures](http://docs.travis-ci.com/user/customizing-the-build/#fast-finishing) would give use more detailed information about which combinations of runtime, environment and exclusions/inclusions do work and which don't, wouldn't it?

(Though, I guess most of the time, this information won't needed to diagnose the issue. So we can leave `fast_finish` on `true`.)

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+-  Django REST Framework (3.2)
+
+Installation
+------------
+
+Install using ``pip``\ …
+
+.. code:: bash
+
+    pip install osmaxx-conversion-service
```
Comment (das-g) [inline README.rst]: Is it a good idea to advertise this, when to use the project, one would also have to install a bunch of non-Python system packages? (See `apt-get install` in `Dockerfile`.)

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    libtool \
+    automake \
+    libprotobuf-c0-dev \
+    protobuf-c-compiler \
+    lua5.2 \
+    liblua5.2-0 \
+    liblua5.2-dev \
+    liblua5.1-0 \
+    \
+    # other dependencies of converters/gis_converter, worker, manager and rest_api
```
Comment (das-g) [inline Dockerfile]: Should we distinguish what dependencies are needed for `converters/gis_converter/` vs. what are needed for `worker`, `manager` and `rest_api`? (We probably shouldn't split them up between the latter three.)

File: README.rst [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+globally, and then simply run:
+
+.. code:: bash
+
+    tox
+
+Documentation
+-------------
+
+To build the documentation, you’ll need to install ``mkdocs``.
```
Comment (das-g) [inline README.rst]: As [MkDocs](http://www.mkdocs.org/) (which is what you get with `pip install mkdocs`) proclaims its purpose to be

> Project documentation with Markdown.

why is this file here in reStructuredText format?

File: docker-compose.yml [hunk 6]
```diff
@@ -1,18 +1,35 @@
-bootstrap:
-  build: bootstrap/
-  links:
-    - database
-  environment:
-    - PGHOST=database
+data:
+  image: busybox
```
Comment (das-g) [inline docker-compose.yml]: Shouldn't we use a base image that is common to our other containers? (probably `debian:jessie`)

File: docker-compose.yml [hunk 7]
```diff
@@ -1,18 +1,35 @@
...
+  command: echo true
   volumes:
-    - ./tmp/osmosis/:/tmp/osmosis/
-extract:
-  build: extract/
+    - /home/njordan/tmp/t:/tmp
```
Comment (das-g) [inline docker-compose.yml]: this doesn't look portable :-1:

File: docs/index.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+$ tox
+```
+
+## Documentation
+
+To build the documentation, you'll need to install `mkdocs`.
+
+```bash
+$ pip install mkdocs
+```
```
Comment (das-g) [inline docs/index.md]: Is ` docs/index.md` created from `README.rst` or the other way around? Or are they independent?

File: .travis.yml (hunk 2 above)
Comment (njordan-hsr) [inline .travis.yml]: I prefer leaving this as is, since at the moment we:
* don't have a buildserver/travis
* only support python 3.4
* make clear, that at a later stage, we would like to move to travis supported testsuite(s).

File: README.rst (hunk 3 above)
Comment (njordan-hsr) [inline README.rst]: Right now, we should probably remove this. But it is quite common, that python packages need some system libraries installed, and then they should be listed here ;-)

File: README.rst (hunk 5 above)
Comment (njordan-hsr) [inline README.rst]: Because it takes the documentation from `docs/*` and not from the Readme.md/.rst.

I'd change that to only use the Readme for a basic introduction, and the `docs/*` for in depth documentation. Currently, they are identical, so merging/joining the Redme.md and the Readme.rst into Readme.md might be a good idea.

File: Dockerfile (hunk 4 above)
Comment (njordan-hsr) [inline Dockerfile]: I think for the moment, this is absolutely sufficient as comment. If we feel that more details are needed, we can do so at any later stage.

File: rest_api/config/settings/production.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
```
Comment (das-g) [inline rest_api/config/settings/production.py]: In Python &ge; 3.0 UTF-8 is the default source encoding, so [PEP 0263](https://www.python.org/dev/peps/pep-0263/)-style magic comments only make sense when a different encoding than that shall be declared.

File: .travis.yml (hunk 2 above)
Comment (das-g) [inline .travis.yml]: :white_check_mark:

File: README.rst (hunk 3 above)
Comment (das-g) [inline README.rst]: :white_check_mark:

File: Dockerfile (hunk 4 above)
Comment (das-g) [inline Dockerfile]: :white_check_mark: let's leave it as-is, then

File: README.rst (hunk 5 above)
Comment (das-g) [inline README.rst]: :white_check_mark:

File: docs/index.md (hunk 8 above)
Comment (das-g) [inline docs/index.md]: Judging from https://github.com/geometalab/osmaxx-postgis-conversion/pull/33#discussion_r43108975, they were independent. With 7b533c6 the `*.rst` is gone, anyway.

:white_check_mark:

File: setup.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
```
Comment (das-g) [inline setup.py]: Is this Python 2?
(There, the encoding declaration would make sense.)

File: test.sh [hunk 11]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+# docker-compose pull
+docker-compose build
+docker-compose run --rm worker1
+#docker-compose run --rm worker 29.525547623634335 40.77546776498174 29.528980851173397 40.77739734768811 -f fgdb -f spatialite -f shp -f gpkg
```
Comment (das-g) [inline test.sh]: Should `test.sh` be checked in?

File: tox.ini [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+[testenv:py34-flake8]
+commands = ./runtests.py --lintonly
+deps =
+       pytest==2.7.0
+       flake8==2.4.0
+
+[testenv:py34-docs]
+commands = mkdocs build
+deps =
+       mkdocs>=0.11.1
```
Comment (das-g) [inline tox.ini]: Doesn't seem to work: Running `tox` (without activated virtualenv) gives me:
```
GLOB sdist-make: /home/das-g/hsr/dev/osmaxx/osmaxx-postgis-conversion/setup.py
py34-flake8 inst-nodeps: /home/das-g/hsr/dev/osmaxx/osmaxx-postgis-conversion/.tox/dist/osmaxx-conversion-service-0.0.1.zip
py34-flake8 installed: flake8==2.4.0,mccabe==0.3.1,osmaxx-conversion-service==0.0.1,pep8==1.5.7,py==1.4.30,pyflakes==0.8.1,pytest==2.7.0,wheel==0.24.0
py34-flake8 runtests: PYTHONHASHSEED='4182549125'
py34-flake8 runtests: commands[0] | ./runtests.py --lintonly
Running flake8 code linting
flake8 passed
py34-docs inst-nodeps: /home/das-g/hsr/dev/osmaxx/osmaxx-postgis-conversion/.tox/dist/osmaxx-conversion-service-0.0.1.zip
py34-docs installed: click==5.1,Jinja2==2.8,livereload==2.4.0,Markdown==2.6.3,MarkupSafe==0.23,mkdocs==0.14.0,osmaxx-conversion-service==0.0.1,PyYAML==3.11,six==1.10.0,tornado==4.2.1,wheel==0.24.0
py34-docs runtests: PYTHONHASHSEED='4182549125'
py34-docs runtests: commands[0] | mkdocs build
INFO    -  Building documentation to directory: /home/das-g/hsr/dev/osmaxx/osmaxx-postgis-conversion/html 
INFO    -  The directory contains stale files. Use --clean to remove them. 
py33-django1.8-drf3.2 create: /home/das-g/hsr/dev/osmaxx/osmaxx-postgis-conversion/.tox/py33-django1.8-drf3.2
ERROR: InterpreterNotFound: python3.3
py34-django1.8-drf3.2 inst-nodeps: /home/das-g/hsr/dev/osmaxx/osmaxx-postgis-conversion/.tox/dist/osmaxx-conversion-service-0.0.1.zip
py34-django1.8-drf3.2 installed: Django==1.8,djangorestframework==3.2.0,osmaxx-conversion-service==0.0.1,py==1.4.30,pytest==2.8.2,pytest-django==2.8.0,wheel==0.24.0
py34-django1.8-drf3.2 runtests: PYTHONHASHSEED='4182549125'
py34-django1.8-drf3.2 runtests: commands[0] | ./runtests.py --fast

==================================== ERRORS ====================================
_________________ ERROR collecting tests/worker/worker_test.py _________________
tests/worker/worker_test.py:5: in <module>
    from worker.converter_job import convert
worker/converter_job.py:3: in <module>
    from django_rq import job
E   ImportError: No module named 'django_rq'
1 error in 0.03 seconds
ERROR: InvocationError: '/home/das-g/hsr/dev/osmaxx/osmaxx-postgis-conversion/runtests.py --fast'
___________________________________ summary ____________________________________
  py34-flake8: commands succeeded
  py34-docs: commands succeeded
ERROR:   py33-django1.8-drf3.2: InterpreterNotFound: python3.3
ERROR:   py34-django1.8-drf3.2: commands failed
```

File: utils/__init__.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+# with statements are use as if they were function, so let's return it as if it was one
```
Comment (das-g) [inline utils/__init__.py]: Is this comment about the snake_case or the 'exported' name vs. the CamelCase of the `import`ed one?

File: utils/__init__.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+from utils.directory_changer_helper import ChDirWith
+
+chg_dir_with = ChDirWith
```
Comment (das-g) [inline utils/__init__.py]: Would it make sense to use the `as` keyword in the `import` instead of making an assignment?

File: utils/directory_changer_helper.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+                        os.chdir(self.new_dir)
+                        self.success = True
+                    except OSError:
+                        if self.fallback_to_tempdir_as_last_resort:
+                            # fake it!
+                            self.new_dir = tempfile.mkdtemp()
+                            sys.stderr.write("Couldn't enter or create directory `"
+                                             + str(self.new_dir) + "'. Entering "
+                                             + "temporary directory instead. All "
+                                             + "data written will be lost!" + '\n')
```
Comment (das-g) [inline utils/directory_changer_helper.py]: Shouldn't we `os.chdir()` into the `self.new_dir` from `tempfile.mkdtemp()`, in this case?

File: converters/osm_cutter.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    p_wget = subprocess.Popen(command.split(), stdout=subprocess.PIPE)
+    command = "osmconvert --out-pbf -o={output_filename} -".format(
+        output_filename=output_filename,
+    )
+    subprocess.check_call(command.split(), stdin=p_wget.stdout)
+    return output_filename
+
+
+GEOMETRY_CLASSES_ACTION = {
+    BBox.__name__: bbox_action,
```
Comment (das-g) [inline converters/osm_cutter.py]: Could we use the class itself instead of its name as dictionary key? As far as I can see, this dictionary or its values don't need to be pickle-able.

File: worker/converter_job.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    :param geometry: osm_cutter.BBox or TBD
+    :param format_type_options: TBD
+    :param output_directory: where results are being stored
+        uses '/tmp/' + time.strftime("%Y-%m-%d_%H%M%S") for default
+    :return: resulting paths/urls for created file
+    """
+
+    # sanity check of input param
+    if geometry.__class__.__name__ not in GEOMETRY_CLASSES_ACTION:
+        raise NotImplementedError(str(type(geometry)) + ' has not been implemented yet')
```
Comment (das-g) [inline worker/converter_job.py]: `str(type(geometry))` is the same as `geometry.__class__.__name__`, isn't it? Which one should we use?

File: .gitignore (hunk 1 above)
Comment (njordan-hsr) [inline .gitignore]: this should now be the case.

File: docker-compose.yml (hunk 6 above)
Comment (njordan-hsr) [inline docker-compose.yml]: Good idea, though this is still ubuntu here (inherited from docker-gdal), I'd like to do that switch on the [243_Implement_PostGIS_Worker branch](https://github.com/geometalab/osmaxx-postgis-conversion/blob/feature/243_Implement_PostGIS_Worker/docker-compose.yml).

File: rest_api/config/settings/production.py (hunk 9 above)
Comment (njordan-hsr) [inline rest_api/config/settings/production.py]: that's true, haven't thought of that - got that line for free ;-)

File: setup.py (hunk 10 above)
Comment (njordan-hsr) [inline setup.py]: No, should actually be python3, corrected.

File: test.sh (hunk 11 above)
Comment (njordan-hsr) [inline test.sh]: At one time, it was useful, but as it stands it does nothing.

File: .gitignore (hunk 1 above)
Comment (das-g) [inline .gitignore]: Is `/html/` the only pattern that should be limited to the root?

File: docker-compose.yml (hunk 6 above)
Comment (das-g) [inline docker-compose.yml]: :white_check_mark: good plan

File: runtests.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
+#! /usr/bin/env python
```
Comment (das-g) [inline runtests.py]: should this be python**3**, too?

File: rest_api/config/settings/production.py (hunk 9 above)
Comment (das-g) [inline rest_api/config/settings/production.py]: :white_check_mark:

File: setup.py (hunk 10 above)
Comment (das-g) [inline setup.py]: :white_check_mark:

File: docker-compose.yml (hunk 7 above)
Comment (das-g) [inline docker-compose.yml]: :white_check_mark: fixed now

File: test.sh (hunk 11 above)
Comment (das-g) [inline test.sh]: :white_check_mark: has been removed

File: .gitignore (hunk 1 above)
Comment (njordan-hsr) [inline .gitignore]: maybe all the `coverage` output as well, but running that inside a docker container might yield the result in a different directory, so I'd just leave it as is, as long as we have nothing colliding with that name we should be OK.

File: tox.ini (hunk 12 above)
Comment (njordan-hsr) [inline tox.ini]: You are missing python 3.3 on your system, it seems. I removed the 3.3 part and brought some other things into position, and works for me (tm).

File: utils/__init__.py (hunk 13 above)
Comment (njordan-hsr) [inline utils/__init__.py]: Yes, the comment is on the wrong line.

File: utils/__init__.py (hunk 14 above)
Comment (njordan-hsr) [inline utils/__init__.py]: let's do it so.

File: worker/converter_job.py (hunk 17 above)
Comment (njordan-hsr) [inline worker/converter_job.py]: they differ for old style classes, can't remember the exact details. Since we're not using old-style classes, we can use either. From a reading perspective, I think `geometry.__class__.__name__` is much more clear on what we are trying to get.

File: converters/osm_cutter.py (hunk 16 above)
Comment (njordan-hsr) [inline converters/osm_cutter.py]: Don't see why we should store classes as keys in a dictionary instead of strings, but other than that I don't see no issue with putting classes there.

File: .gitignore (hunk 1 above)
Comment (das-g) [inline .gitignore]: The venv folder names are more prone to collisions, I think.

:white_check_mark: but let's leave it as-is for now

File: tox.ini (hunk 12 above)
Comment (das-g) [inline tox.ini]: > You are missing python 3.3 on your system, it seems.

Yeah, that's what

    ERROR:   py33-django1.8-drf3.2: InterpreterNotFound: python3.3
in the summary output is about. But 3.4 seems to have failed, too:

    ERROR:   py34-django1.8-drf3.2: commands failed

I'll try again with your changes.

File: utils/__init__.py (hunk 13 above)
Comment (das-g) [inline utils/__init__.py]: :white_check_mark:  I've rephrased the comment: 53404d5

File: utils/__init__.py (hunk 14 above)
Comment (das-g) [inline utils/__init__.py]: Hu. This causes a warning in PyCharm, whereas the assignment didn't. :stuck_out_tongue_winking_eye:

:white_check_mark: I'd leave it as it is now, anyway, though.

File: worker/converter_job.py (hunk 17 above)
Comment (das-g) [inline worker/converter_job.py]: :white_check_mark: now `.__class__` is used in all occurrences.

File: converters/osm_cutter.py (hunk 16 above)
Comment (das-g) [inline converters/osm_cutter.py]: :white_check_mark: I see you've changed it, regardless: 3b3522e

The reason why I prefer classes as keys here, is that the decision how to cut out a boundary should be based on the class identity, not the class name. This way, our decision logic would keep working even if we decided to name several bounding geometry classes the same and just have them in separate modules or packages. (Which, granted, wouldn't be a good idea for readability reasons.)

File: tox.ini [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+envlist =
+       py34-{flake8,docs},
+       {py34}-django{1.8}-drf{3.2}
+
+[testenv]
+commands = ./runtests.py --fast
+setenv =
+       PYTHONDONTWRITEBYTECODE=1
+deps =
+       -rrequirements.txt
```
Comment (das-g) [inline tox.ini]: Could this be written with a space? (`-r requirements.txt`)

File: pylintrc [hunk 20]
```diff
@@ -0,0 +1,384 @@
+[MASTER]
```
Comment (das-g) [inline pylintrc]: Has this file been [generated](http://docs.pylint.org/faq.html?highlight=pylintrc#how-do-i-find-the-option-name-for-pylintrc-corresponding-to-a-specific-command-line-option)? If so:
* What options have been used for generation?
* Has the file been manually modified after generation, and if so, how?

File: tox.ini (hunk 12 above)
Comment (das-g) [inline tox.ini]: :white_check_mark: Yay, works now.

File: tox.ini (hunk 19 above)
Comment (njordan-hsr) [inline tox.ini]: No.

File: pylintrc (hunk 20 above)
Comment (njordan-hsr) [inline pylintrc]: It has been generated, and used the default values (ie. no extra options). Since I don't know which of the gazillion values we should check for, and not even which ones could possibly conflict with Django best practices, I'm abandoning this for now. If you're interested in doing this, go for it. For now, we can still run this as an _advisory_ checker. I'm very strongly against using this, since it's correct usage requires a lot of time that is better spent somewhere else.

File: runtests.py (hunk 18 above)
Comment (das-g) [inline runtests.py]: :white_check_mark: has been changed

File: utils/directory_changer_helper.py (hunk 15 above)
Comment (das-g) [inline utils/directory_changer_helper.py]: :white_check_mark: codepath culled in de586db

File: pylintrc (hunk 20 above)
Comment (das-g) [inline pylintrc]: :white_check_mark: fine with me
